People: Accept invite by email subscription only

roccotripaldi opened the pull request:
closes #668 

This PR adds a footer link to the logged out invite accept, allowing an invited user to follow a blog by email subscription only.

To Test
1. Invite a non-wpcom email address to follow your blog
2. log out of wordpress
3. go to /accept-invite/$blog/$inviteKey/$activation_key/$management_key
4. click the link at the bottom reading 'Or follow by email subscription only.'
5. Ensure that you are now following the blog and are redirected to subscribe.wordpress.com

cc: @lezama @ebinnion

roccotripaldi commented:
Note: we'll need to add the `subscriptionActivationKey` to _all_ follower requests. but that can be handled in another PR.

lezama commented on client/my-sites/invites/index.js:
```diff
@@ -6,11 +6,12 @@ import page from 'page';
 /**
  * Internal dependencies
  */
-import { acceptInvite } from './controller';
```
Personally I find it more beautiful :relaxed: if we do `{ acceptInvite, saveSubscriptionActivationKey }` instead of importing `controller`

lezama commented on client/my-sites/invites/controller.js:
```diff
@@ -9,15 +9,32 @@ import React from 'react';
+
+export default {
+	saveSubscriptionActivationKey( context, next ) {
+		if ( context.query.activate ) {
```
can't we add the activation key to the path instead of passing it as a query?

lezama commented on client/my-sites/invites/invite-accept-logged-out/index.jsx:
```diff
@@ -43,13 +43,13 @@ export default React.createClass( {
 	},
 
-	getFormHeader() {
+	renderFormHeader() {
```
I like it

lezama commented on client/my-sites/invites/invite-accept-logged-out/index.jsx:
```diff
@@ -43,13 +43,13 @@ export default React.createClass( {
 		);
 	},
 
-	loginUser() {
```
I choose this name because it is more like an action, as the form will never get shown, this will render the form and autosubmit it in order to login the user. I don't mind changing it if you feel it makes more sense :)

lezama commented on client/my-sites/invites/invite-accept-logged-out/index.jsx:
```diff
@@ -60,30 +60,61 @@ export default React.createClass( {
-		let logInUrl = config( 'login_url' ) + '?redirect_to=' + encodeURIComponent( window.location.href );
+	subscribeUserByEmailOnly() {
+		this.setState( { submitting: true } );
+		acceptInvite(
```
I think we should change what `acceptInvite` receives to a single object with multiple keys, what do you think?

lezama commented:
This is looking good :)

roccotripaldi commented:
@lezama  thanks for the review. I've address all your feedback. I'm going to look into changing the redirect, and try to do it in this PR. 

To clarify, we should redirect to susbscribe.wordpress.com/?update=activated&email=$email&key=$key
with YET ANOTHER key. 

I think it may be possible to return the latest key on the API after the subscription is activate.

roccotripaldi commented:
I changed my mind. In keeping more with the current way we do things, I decided to add the 3rd key to the URL. so now we'll have
/accept-invite/$site/$invite_key/$activation_key/$management_key for follower invites.

All other invites will only have $invite_key

lezama commented on client/my-sites/invites/controller.js:
```diff
@@ -10,14 +10,19 @@ import i18n from 'lib/mixins/i18n';
 
-export function acceptInvite( context ) {
-	titleActions.setTitle( i18n.translate( 'Accept Invite', { textOnly: true } ) );
+export default {
```
this could be `export function acceptInvite(`

lezama commented on client/my-sites/invites/invite-accept-logged-out/index.jsx:
```diff
@@ -60,27 +60,56 @@ export default React.createClass( {
+	subscribeUserByEmailOnly() {
+		this.setState( { submitting: true } );
+		acceptInvite(
+			Object.assign( this.props.invite, { activation_key: this.props.subscriptionActivationKey } ),
```
Why do we need to do this?

roccotripaldi replied:
the activation key?

lezama replied:
:+1: I had misread the code here :)

lezama commented on client/my-sites/invites/invite-accept-logged-out/index.jsx:
```diff
@@ -60,27 +60,56 @@ export default React.createClass( {
+		return (
+			<div>
+				<a href={ logInUrl } className="logged-out-form__link">
+					{ this.translate( 'Already have a WordPress.com account? Log in now.' ) }
```
I think both links should be in the same translation string in order to give Translators some context. "Or" could be tricky to place/translate.

lezama commented on client/my-sites/invites/invite-accept/index.jsx:
```diff
@@ -82,8 +82,14 @@ export default React.createClass( {
+			: <LoggedOut
+				{ ...this.state.invite }
+				redirectTo={ this.getRedirectTo() }
+				subscriptionActivationKey={ this.props.activation_key }
```
Maybe we should put this properties inside the invite object?

roccotripaldi replied:
Should i do that here in Calypso, or should we see if it is possible to do it on the API?

lezama replied:
I meant here, but now that you mention, can't we get rid of at least one of the 3 keys?

roccotripaldi commented:
@lezama - i did an other iteration taking your feedback into account.

For now, because of API limitations, I'm keeping all 3 keys in the URL

In another PR, we can clean up the URL by perhaps doing something like the signup flow is using,
where it accepts URL GET parameters, adds them to the component, and then makes the URL look pretty by striping out the parameters.

ebinnion commented:
Perhaps i'm misunderstand, but an API limitation doesn't necessarily impact whether we do `/accept-invite/$site/$invite_key/$action` or `/accept-invite/$site/$invite_key?activation`. We just need to pass the correct data to the API.

I'm fine with either approach, just wanted to point that out.

ebinnion commented on client/my-sites/invites/controller.js:
```diff
@@ -17,7 +17,10 @@ export function acceptInvite( context ) {
-		React.createElement( InviteAccept, context.params ),
+		React.createElement(
+			InviteAccept,
+			context.params
```
My guess is that if we moved to using `GET` parameters, the heavy lifting would be done here. In this method we would get the `GET` parameters above (See what I did there :stuck_out_tongue: :eyeglasses: ) and then we would do the `page.replace()` or similar.

The only other change I think we would need is updating the route in `my-sites/invites/index.js` to be `'/accept-invite/:site_id/:invitation_key`.

ebinnion commented on client/my-sites/invites/invite-accept-logged-out/index.jsx:
```diff
@@ -60,11 +60,40 @@ export default React.createClass( {
+				if ( error ) {
+					this.setState( { error } );
+				} else {
+					window.location = 'https://subscribe.wordpress.com?update=activate&email=' + encodeURIComponent( this.props.invite.meta.sent_to ) + '&key=' + this.props.invite.authKey;
```
Why do we need to send the user to `subscribe.wordpress.com` if we're using the API to accept the invite?

roccotripaldi replied:
I like that idea! I was just mimicking the current flow.

ebinnion commented on client/my-sites/invites/invite-accept/index.jsx:
```diff
@@ -46,6 +46,11 @@ export default React.createClass( {
 	refreshInvite() {
 		const invite = InvitesStore.getInvite( this.props.site_id, this.props.invitation_key );
 		const error = InvitesStore.getInviteError( this.props.site_id, this.props.invitation_key );
+		// add subscription-related keys to the invite
```
There should be an empty newline above this comment.

ebinnion commented on client/my-sites/invites/invite-accept/index.jsx:
```diff
@@ -46,6 +46,11 @@ export default React.createClass( {
+		Object.assign( invite.invite, {
+			activationKey: this.props.activation_key,
+			authKey: this.props.auth_key
+		} );
```
I was wondering if you had change the endpoint to add the `activationKey`. But, now I see where you added it :smile:

ebinnion commented on client/my-sites/invites/invite-accept/index.jsx:
```diff
@@ -87,7 +92,8 @@ export default React.createClass( {
 		return user
-			? <LoggedIn { ...this.state.invite } decline={ this.decline } user={ user } />
+			? <LoggedIn
+				{ ...this.state.invite } decline={ this.decline } user={ user } />
```
Any reason why we dropped these props to the next line?

ebinnion commented on client/my-sites/invites/invite-accept-logged-out/index.jsx:
```diff
@@ -60,11 +60,40 @@ export default React.createClass( {
+	},
+
+	renderEmailOnlySubscriptionLink() {
+		if ( this.props.invite.meta.role !== 'follower' || ! this.props.invite.activationKey ) {
```
At first glance, I thought this could prevent a potential issue with `viewers` on private WPCOM sites since `viewers` are also labeled as `followers` in the code.

But, since we control the invitation email, and since we don't need an activation key for viewers, I don't anticipate this being an issue.

ebinnion commented:
Functionally, this seemed to work well. The only issue I noticed is that after clicking the "Follow by email subscription only" link, there was a lag of a second or two before I was redirected without the UI updating.

Perhaps, we could do something similar to #1237 and disable the form once that link is clicked?

roccotripaldi commented:
Thanks for the review Eric. I addressed all your feedback.

I was already doing `setState( { submitting: true} );` when we clicked the email only link, but the big blue buttons wasn't been disabled. I fixed that here:

https://github.com/Automattic/wp-calypso/pull/1190/files#diff-c1058d9fea1b7d007377a2a1e78a3cccR377

ebinnion commented:
The disabling looks good to me!

I am getting a JS error after accepting an invite and then going back to `/accept-invite/$invite_key/$activation/$manage`.

![Uploading Screen Shot.png…]()

My guess is that it's related to [this block in refreshInvite()](https://github.com/Automattic/wp-calypso/blob/accept-invite-email-only/client/my-sites/invites/invite-accept/index.jsx#L50-L54). We should probably wrap that in an:

```js
if ( ! error ) {
  // Do work
}  
```
